src/reader/npz_reader.py

```diff
@@ -40,48 +41,81 @@ def read(self, epoch_number):
+
             if self.sample_shuffle != Shuffle.OFF:
                 if self.sample_shuffle == Shuffle.SEED:
                     random.seed(self.seed)
-                random.shuffle(num_sets)
-            for num_set in num_sets:
-                # Should we check if profiling is enabled and if we're using PT?
-                with tf.profiler.experimental.Trace('HDF5 Input', step_num=num_set / self.batch_size, _r=1):
-                    progress(count, total, "Reading NPZ Data")
-                    count += 1
-                    images = element['dataset'][:][:][num_set * self.batch_size:(num_set + 1) * self.batch_size - 1]
-                yield images
+                random.shuffle(sample_index_list)
+            for sample_index in sample_index_list:
+                count += 1
+                logging.info(f"{utcnow()} num_set {sample_index} current batch_size {len(batch)}")
+                my_image = self._dataset[index]['data'][..., sample_index]
+                logging.debug(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+
+                my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+                logging.debug(f"{utcnow()} new shape of image {my_image_resized.shape}")
+                batch.append(my_image_resized)
+                is_last = 0 if count < total else 1
+                if is_last:
+                    while len(batch) is not self.batch_size:
+                        batch.append(np.random.rand(self.max_dimension, self.max_dimension))
+                if len(batch) == self.batch_size:
+                    batch = np.array(batch)
+                    yield is_last, batch
+                    batch = []
+
+    def read_index(self, index):
+        file_index = math.floor(index / self.num_samples)
+        if self._dataset[file_index]["data"] is None:
+            with np.load(self._dataset[file_index]["file"], allow_pickle=True) as data:
+                self._dataset[file_index]['data'] = data["x"]
+        element_index = index % self.num_samples
+        my_image = self._dataset[file_index]['data'][..., element_index]
+        logging.info(f"{utcnow()} shape of image {my_image.shape} self.max_dimension {self.max_dimension}")
+        my_image_resized = np.resize(my_image, (self.max_dimension, self.max_dimension))
+        logging.info(f"{utcnow()} new shape of image {my_image_resized.shape}")
+        return my_image_resized
 
-    def finalize(self):
-        pass
+    def get_sample_len(self):
+        total_samples = 0
+        for index in range(len(self._dataset)):
+            if self._dataset[index]["data"] is None:
+                with np.load(self._dataset[index]["file"], allow_pickle=True) as data:
+                    self._dataset[index]['data'] = data["x"]
+            total_samples = total_samples + self._dataset[index]['data'].shape[2]
+        return total_samples
```

Comment:
This is too expensive; it essentially read all the dataset. If the dataset size is large, it will have trouble. Let us use the same way you used for tfreader?

Reply:
Fixed